posts/benchmarking_design_decisions/index.qmd

```diff
@@ -0,0 +1,228 @@
+
+```{r}
+try(rnorm())
+```
+
+Warnings are used to handle less fatal issues with code execution that do not
+need to halt the process.
+
+```{r}
+c(1:2) > c(1:3)
+```
+
+Base R ships with functionalities for condition handling and signalling. There is
+`stop()` for signaling errors and `warning()` for throwing warnings. These base R
+functions are, however, are not flexible enough for advanced formatting: colour,
+progress bar, contextual pluralisation, and so forth.
+
+Package developers often have to adopt external dependencies for flexibility
+in condition signalling. The [cli](https://cli.r-lib.org/articles/semantic-cli.html) package is well designed for this. `cli` offers many advanced features
+that will not be discussed here, but of note is its `cli_abort()` and
+`cli_warn()` functions for signalling errors and warnings respectfully.
+
+`cli` is a great candidate for flexibility in condition signalling but one
+might naturally wonder if this comes at a cost. Here,
+we will investigate this through a benchmarking exercise against base R's
+`warning()` function. The results of this exercise will provide us with insights
+on whether its worth taking on `cli` for its flexibility as a trade-off for its
+slower speed.
+
+## Experiment
+
+We will consider different scenarios of number of warnings thrown to
+tease out the relationship between speed of the function and number of warnings
+thrown in a function with and without `cli::cli_warn()`. We will also check if
+the benchmark scales with the number of warnings in the function.
+
+Based on this description, we will need the following packages: `cli` for its warning function,
+[bench](https://cran.r-project.org/web/packages/bench/index.html) for measuring
+the run times, `purrr`, `tidyr`, and `dplyr` for running the functions and
+wrangling the results, and and `ggplot2` for visualizing the results.
+
+```{r}
+library(bench)
+library(cli)
+library(ggplot2)
+library(purrr)
+library(dplyr)
+library(tidyr)
+```
+
+Let's define a function that takes an argument `n` for the number of warnings
+to throw and `pkg` for the type of package to use.
+
+```{r}
+warn_Ntimes <- function(n, pkg) {
+  warning_msg <- "x must be an integer"
+  switch(pkg,
+    base = for (i in 1:n) {
+      warning(warning_msg)
+    },
+    cli = for (i in 1:n) {
+      cli_warn(warning_msg)
+    }
+  )
+}
+```
+
+Let's test our function to see if it works as expected.
+
+```{r}
+warn_Ntimes(3, "base")
+warn_Ntimes(3, "cli")
+```
+
+Now, we'll consider scenarios where a function throws 1, 5, 10, 15, 20, and
+100 warnings using base R and `cli`.
+
+```{r warning=FALSE}
+# Number of warnings to throw
+warnings <- c(1, 5, 10, 15, 20, 100)
+```
+
+Let's run benchmarks over the different scenarios and store results
+in a data.frame.
+
+```{r warning=FALSE}
+bnmark_list <- warnings |>
+  map(
+    \(x) mark(
+      "cli::cli_warn()" = warn_Ntimes(x, pkg = "cli"),
+      "base::warning()" = warn_Ntimes(x, pkg = "base")
+    )
+  )
```

Comment:
Since you're using bench, it's probably simpler to use `press()` directly. It should be something like:

```r
bench::press(
  fct = c("cli", "base"),
  times = c(1, 5, 10, 15, 20, 100),
  warn_Ntimes(times, fct)
)
```

https://bench.r-lib.org/reference/press.html

Reply:
Wow, that's very convenient. Thanks.